pkg/volume/aws_ebs/aws_util.go

```diff
@@ -112,62 +148,76 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner, node *
-		if node.ObjectMeta.Labels == nil {
-			return "", 0, nil, "", fmt.Errorf("labels for node missing")
-		}
-		z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
-		if !ok {
-			return "", 0, nil, "", fmt.Errorf("%s Label for node missing", kubeletapis.LabelZoneFailureDomain)
-		}
-
-		if volumeOptions.ZonePresent || volumeOptions.ZonesPresent {
-			return "", 0, nil, "", fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode set to WaitForFirstConsumer.")
-		}
-
-		volumeOptions.ZonePresent = true
-		volumeOptions.AvailabilityZone = z
-	}
-
-	// TODO: implement PVC.Selector parsing
-	if c.options.PVC.Spec.Selector != nil {
-		return "", 0, nil, "", fmt.Errorf("claim.Spec.Selector is not supported for dynamic provisioning on AWS")
-	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) {
+		if node != nil {
+			// if node set, zone/zones cannot be specified as node's zone may conflict
+			if volumeOptions.ZonePresent || volumeOptions.ZonesPresent {
+				return nil, fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode set to WaitForFirstConsumer.")
+			}
 
-	name, err := cloud.CreateDisk(volumeOptions)
-	if err != nil {
-		glog.V(2).Infof("Error creating EBS Disk volume: %v", err)
-		return "", 0, nil, "", err
-	}
-	glog.V(2).Infof("Successfully created EBS Disk volume %s", name)
+			// if node is set, we pick node's zone and ignore any allowedTopologies (since scheduler has accounted for it already)
+			if node.ObjectMeta.Labels == nil {
+				return nil, fmt.Errorf("labels for node missing")
+			}
 
-	labels, err := cloud.GetVolumeLabels(name)
-	if err != nil {
-		// We don't really want to leak the volume here...
-		glog.Errorf("error building labels for new EBS volume %q: %v", name, err)
+			z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+			if !ok {
+				return nil, fmt.Errorf("%s Label for node missing", kubeletapis.LabelZoneFailureDomain)
+			}
+			volumeOptions.ZonePresent = true
+			volumeOptions.AvailabilityZone = z
+		} else {
+			// node not specified so collect all specified zones from allowedTopologies
+			allowedZones := make([]string, 0)
+			for _, term := range allowedTopologies {
+				zoneRequirementFound := false
+				for _, exp := range term.MatchLabelExpressions {
+					if exp.Key == kubeletapis.LabelZoneFailureDomain {
+						if zoneRequirementFound {
+							// A TopologySelectorTerm in allowedTopologies should NOT have  multiple TopologySelectorLabelRequirement with LabelZoneFailureDomain key
+							// since expressions with LabelZoneFailureDomain cannot be "and" ed for EBS; only "or" ed
+							return nil, fmt.Errorf("multiple expressions with key %s cannot be specified in allowedTopology %v", exp.Key, term)
+						}
+						for _, value := range exp.Values {
+							allowedZones = append(allowedZones, value)
```

Comment:
Should this be a set?  The same zone could have mistakenly been specified multiple times.

Reply:
We can test with some examples.  But if it doesn't validate, we'll have to think about if we can strengthen it; the validation reuses much of the existing selector validation, and we can't tighten that.

Reply:
> the validation reuses much of the existing selector validation, and we can't tighten that.

If it does turn out that current validation isn't sufficient, then I'd argue either the existing selector validation is somehow not strong enough, or maybe we shouldn't use these selectors

Reply:
Can you open a bug for that? We can look into fixing that separately

Reply:
Looked at the current validation code briefly and it doesn't seem like we have any structural validation today.

I'm happy to take this task if folks agree this is a good idea: https://github.com/kubernetes/kubernetes/issues/66184